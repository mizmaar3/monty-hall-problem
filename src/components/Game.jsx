import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { Button, Badge } from "react-bootstrap";
import Door from "containers/door";
import Status from "containers/status";
import { WON, LOST } from "constants/status";
import { AWARD, TROLL } from "constants/content";

const Game = props => {
  const { data, changeSelectionAfterTrick } = props;
  const dataIndices = data.map((v, i) => i);

  const [status, setStatus] = useState("");
  const [trivia, setTrivia] = useState("");
  const [selected, setSelected] = useState([]);
  const [opened, setOpened] = useState([]);
  const [onSelectionText, setOnSelectionText] = useState(false);
  const [onOpenText, setOnOpenText] = useState(false);
  const [selectionChangeOptionText, setSelectionChangeOptionText] = useState(
    false
  );

  const closedAndNotSelected = () =>
    dataIndices.find(v => v !== selected[0] && !opened.includes(v));

  const onSelection = e => {
    const selection = +e.target.value;
    if (changeSelectionAfterTrick) {
      setSelected([selection]);
    } else {
      setOpened([selection]);
      setStatus(data[selection] === 1 ? WON : LOST);
    }
  };

  const trickThePlayer = () => {
    const doorToOpen = dataIndices.find(
      v => v !== selected[0] && data[v] !== 1
    );
    setOpened([doorToOpen]);
  };

  const conclude = (selection, doorsToOpen, didChange = false) => {
    setOpened(doorsToOpen);
    setStatus(selection === 1 ? WON : LOST);
    setSelectionChangeOptionText(false);
    let triviaText = "Wow nice luck!";
    if (didChange) {
      triviaText =
        selection === 1
          ? "Good that you changed your selection :)"
          : "Hard luck!";
    } else {
      triviaText =
        selection === 1
          ? "Wow nice luck!"
          : "You had a chance to change your selection!";
    }
    setTrivia(triviaText);
  };

  const onChangeSelection = () => {
    const changedSelection = closedAndNotSelected();
    const doorsToOpen = [...opened, changedSelection];
    conclude(changedSelection, doorsToOpen, true);
  };

  const onKeepSelection = () => {
    const doorsToOpen = [...opened, ...selected];
    conclude(data[selected[0]], doorsToOpen, false);
  };

  useEffect(() => {
    if (selected.length === 1) {
      trickThePlayer();
      setOnSelectionText(true);
    }
  }, [selected]);

  useEffect(() => {
    if (opened.length === 1 && changeSelectionAfterTrick) {
      setOnOpenText(true);
    }
  }, [opened]);

  useEffect(() => {
    if (opened.length === 1 && changeSelectionAfterTrick) {
      setSelectionChangeOptionText(true);
    }
  }, [opened]);

  return (
    <div
      style={{
        width: "80%",
        pointerEvents: status ? "none" : "auto",
        position: "relative",
        padding: "10px",
        fontSize: "14px",
      }}
    >
      <div
        style={{
          display: "flex",
          justifyContent: "space-between",
          marginBottom: "10px",
        }}
      >
        {data.map((d, i) => {
          return (
            <div key={`door-${i}`} style={{ height: "fit-content" }}>
              <Door
                open={opened.includes(i)}
                selected={selected[0] === i}
                content={d === 1 ? AWARD : TROLL}
              />
              {selected.length === 0 && (
                <Button
                  style={{
                    marginTop: "10px",
                  }}
                  variant="primary"
                  value={i}
                  onClick={onSelection}
                >
                  Select Door {i + 1}
                </Button>
              )}
            </div>
          );
        })}
      </div>

      {onSelectionText && (
        <div>
          You selected <Badge variant="primary">Door {+selected[0] + 1}</Badge>
        </div>
      )}

      {onOpenText && (
        <div>
          The computer opened{" "}
          <Badge variant="info">Door {+opened[0] + 1}</Badge> and it was a GOAT
        </div>
      )}

      {selectionChangeOptionText && (
        <div>
          You can change your selection from{" "}
          <Badge variant="secondary">Door {+selected[0] + 1} </Badge> to{" "}
          <Badge variant="secondary">Door {closedAndNotSelected() + 1}</Badge>
          <br />
          <br />
          <Button variant="primary" size={"sm"} onClick={onChangeSelection}>
            Yes change selection
          </Button>{" "}
          <Button variant="info" size={"sm"} onClick={onKeepSelection}>
            Keep selection and open the door
          </Button>
        </div>
      )}

      {trivia && (
        <div>
          <strong>{trivia}</strong>
        </div>
      )}

      {status && <Status content={status} />}
    </div>
  );
};

Game.defaultProps = {
  changeSelectionAfterTrick: true,
};

Game.propTypes = {
  data: PropTypes.array.isRequired,
  changeSelectionAfterTrick: PropTypes.bool,
};

export default Game;
