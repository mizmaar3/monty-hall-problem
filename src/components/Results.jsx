import React from "react";
import { Spinner, Card, Badge } from "react-bootstrap";
import { useSelector } from "react-redux";

const Results = () => {
  const {
    isFetching,
    items: { cars, goats, size, changeSelection },
  } = useSelector(state => state.gameSimulation);

  const carPercentage = Math.round((cars / size) * 100);
  const goatsPercentage = 100 - carPercentage;

  return isFetching ? (
    <Spinner animation="border" variant="primary" />
  ) : size > 0 ? (
    <Card>
      <Card.Header>
        <strong>
          Simulation result for{" "}
          <Badge pill={true} variant="warning">
            {size}
          </Badge>{" "}
          tries
        </strong>
      </Card.Header>
      <Card.Body>
        <div style={{ display: "flex", flexDirection: "column" }}>
          <div
            style={{
              display: "flex",
              height: "40px",
              margin: "5px 0",
              flexDirection: "row",
              color: "#fff",
            }}
          >
            <div
              style={{
                width: `${carPercentage}%`,
                height: "100%",
                background: "#28a745",
                alignItems: "center",
                justifyContent: "center",
                display: "flex",
              }}
            >
              {carPercentage}%
            </div>
            <div
              style={{
                width: `${goatsPercentage}%`,
                height: "100%",
                background: "#dc3545",
                alignItems: "center",
                justifyContent: "center",
                display: "flex",
              }}
            >
              {goatsPercentage}%
            </div>
          </div>
          <div
            style={{
              display: "flex",
              flexDirection: "row",
              justifyContent: "space-between",
              fontSize: "12px",
              fontWeight: "bold",
            }}
          >
            <div>No. of Cars: {cars}</div>
            <div>No. of Goats: {goats}</div>
          </div>
        </div>
      </Card.Body>
      <Card.Footer>
        Change selection was{" "}
        <strong>{changeSelection ? "Enabled" : "Disabled"}</strong>
      </Card.Footer>
    </Card>
  ) : null;
};

export default Results;
