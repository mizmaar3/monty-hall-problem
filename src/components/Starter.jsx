import React, { useEffect, useState, useRef } from "react";
import { Card, Button, InputGroup, Form } from "react-bootstrap";
import { useDispatch, useSelector } from "react-redux";
import { getGameData } from "actions/gameData";
import { gameSimulation, resetToInitial } from "actions/gameSimulation";
import { setFormat } from "actions/format";
import { GAME, SIMULATION } from "constants/format";

const Starter = () => {
  const dispatch = useDispatch();
  const [noOfSimulation, setNoOfSimulation] = useState(0);
  const [selectionChangeOnTrick, setSelectionChangeOnTrick] = useState(true);
  const noOfSimulationRef = useRef(null);
  const selectionChangeInputRef = useRef(null);
  const doorSelectionRef = useRef(null);

  const { isFetching: loading } = useSelector(state => state.gameSimulation);

  useEffect(() => {
    if (noOfSimulation > 0 && loading) {
      const selection = doorSelectionRef.current.value;
      const changeSelection = selectionChangeInputRef.current.checked;
      setTimeout(() => {
        dispatch(gameSimulation(noOfSimulation, selection, changeSelection));
      }, 100);
    }
  }, [noOfSimulation, loading]);

  const noop = () => 0;

  const onLoadSimulation = () => {
    dispatch(resetToInitial());
    dispatch(setFormat(SIMULATION));
    const noOfSimulation = +noOfSimulationRef.current.value;
    setNoOfSimulation(noOfSimulation);
  };

  const onPlayGame = () => {
    dispatch(getGameData(1));
    dispatch(setFormat(GAME));
  };

  const onCheckBoxChange = e => setSelectionChangeOnTrick(e.target.checked);

  return (
    <Card className={"my-auto"}>
      <Card.Header>Let's Make a Deal</Card.Header>
      <Card.Body>
        <Card.Text>
          You will see three closed doors - behind one is a car, and behind the
          other two are goats. Select one door and try your luck. The computer
          might trick you. Be prepared!
        </Card.Text>
        <br />
        <InputGroup className="mb-3">
          <InputGroup.Prepend>
            <InputGroup.Text>Select Number of Simulation</InputGroup.Text>
          </InputGroup.Prepend>
          <Form.Control as="select" ref={noOfSimulationRef}>
            <option>10</option>
            <option>50</option>
            <option>100</option>
            <option>1000</option>
            <option>5000</option>
          </Form.Control>
        </InputGroup>
        <InputGroup className="mb-3">
          <InputGroup.Prepend>
            <InputGroup.Text>Select Door Number</InputGroup.Text>
          </InputGroup.Prepend>
          <Form.Control as="select" ref={doorSelectionRef}>
            <option value="0">Door 1</option>
            <option value="1">Door 2</option>
            <option value="2">Door 3</option>
          </Form.Control>
        </InputGroup>
        <div className="mb-3">
          <Form.Check
            ref={selectionChangeInputRef}
            type="checkbox"
            checked={selectionChangeOnTrick}
            onChange={onCheckBoxChange}
            label="Enable selection change after computer tricks you"
          />
        </div>
        <Button
          variant="primary"
          onClick={loading ? noop : onLoadSimulation}
          disabled={loading}
        >
          Load Simulation
        </Button>{" "}
        <Button
          variant="secondary"
          onClick={loading ? noop : onPlayGame}
          disabled={loading}
        >
          Play Game Instead
        </Button>
      </Card.Body>
    </Card>
  );
};

export default Starter;
