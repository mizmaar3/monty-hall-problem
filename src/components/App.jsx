import React from "react";
import { useSelector } from "react-redux";
import Game from "./Game";
import Starter from "./Starter";
import Results from "./Results";
import { GAME } from "constants/format";

const App = () => {
  const {
    gameData: {
      items: { data: gameData },
    },
    format,
  } = useSelector(state => state);

  const renderGame = format === GAME && gameData[0] && gameData[0].length > 0;

  return renderGame ? (
    <Game data={gameData[0]} changeSelectionAfterTrick={true} />
  ) : (
    <>
      <Starter />
      <br />
      <Results />
    </>
  );
};

export default App;
