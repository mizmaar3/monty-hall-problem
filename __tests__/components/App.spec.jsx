import React from "react";
import { shallow } from "enzyme";
import App from "components/App";
import { useSelector } from "react-redux";
import Starter from "components/Starter";
import Game from "components/Game";
import Results from "components/Results";

jest.mock("components/Starter");
Starter.mockImplementation(() => <div>starter</div>);

jest.mock("components/Game");
Game.mockImplementation(() => <div>game</div>);

jest.mock("components/Results");
Results.mockImplementation(() => <div>results</div>);

jest.mock("react-redux");

describe("App component", () => {
  it("Should render <Starter> when load", () => {
    useSelector.mockImplementation(() => ({
      format: "",
      gameData: {
        items: {
          data: [],
        },
      },
    }));
    const wrapper = shallow(<App />);
    expect(wrapper.html().includes("<div>starter</div>")).toBe(true);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it("Should render Game component when there is game data", () => {
    useSelector.mockImplementation(() => ({
      format: "GAME",
      gameData: {
        items: {
          data: [[1, 0, 0]],
        },
      },
    }));
    const wrapper = shallow(<App />);
    expect(wrapper.html().includes("<div>game</div>")).toBe(true);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });
});
