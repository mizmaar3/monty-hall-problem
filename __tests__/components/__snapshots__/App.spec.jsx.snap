// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component Should render <Starter> when load 1`] = `
<Fragment>
  <Starter />
  <br />
  <Results />
</Fragment>
`;

exports[`App component Should render Game component when there is game data 1`] = `
<Game
  changeSelectionAfterTrick={true}
  data={
    Array [
      1,
      0,
      0,
    ]
  }
/>
`;
